src/librustc_typeck/astconv.rs

```diff
@@ -706,10 +707,16 @@ impl<'o, 'gcx: 'tcx, 'tcx> AstConv<'gcx, 'tcx>+'o {
                         .emit();
         }
 
+        // Dedup auto traits so that `dyn Trait + Send + Send` is the same as `dyn Trait + Send`.
+        let mut auto_traits =
+            auto_traits.into_iter().map(ty::ExistentialPredicate::AutoTrait).collect::<Vec<_>>();
+        auto_traits.sort_by(|a, b| a.cmp(tcx, b));
+        auto_traits.dedup_by(|a, b| (&*a).cmp(tcx, b) == Ordering::Equal);
```

Comment:
The (&*a) is necessary here because `std::cmp::Ord::cmp` shadows `ty::ExistentialPredicate::cmp` for an `&mut ty::ExistentialPredicate`, but not necessary in the `sort_by` on the previous line because the shadowing is reversed for `&ty::ExistentialPreciate` (and the unborrowed version).

Reply:
I wonder if it's possible for you to just collect into an `FxHashSet` here?

Reply:
The `cmp` used by a HashSet doesn't compare using the TyCtx, so would fail if the DefIds for the auto trait are different, which I think can happen through type or use aliases. Also, in the vast majority of cases, there's going to be less than three auto traits (and many times zero), so I don't think it would be much of an optimization.

Reply:
First off, I would recommend invoking `ty::ExistentialPredicate::cmp` — but maybe we should rename that method, if it is indeed not equivalent to `Ord::cmp`?

Reply:
Put another way, why can't we invoke `.dedup()` here?

Reply:
Hmm, I thought that the `DefIds` might be different based on which `use` clause you use, but switching to just `.dedup()` works here. But then, why is there both `ty::ExistentialPredicate::cmp` and `Ord::cmp` if they give the same results? Are there actually cases where two `DefId`s could point to the same trait?

Reply:
So, that function (a) has a bad name and (b) needs a comment. I think the distinction is this:

- `ExistentialPredicate::cmp` sorts via a **stable ordering** that will not change if modules are ordered or other changes are made to the tree. In particular, this ordering is preserved across incremental compilations.
- `Ord::Cmp` sorts in an order that is not stable across compilations.

For purposes of equality, they are equivalent.

It should really be called `ExistentialPredicate::stable_cmp` -- would you be game to try and rename it? (And add a comment.)

That could be either part of this PR or outside of it...

...given the unfortunate shadowing, the way I would go about it would be to add `stable_cmp` and mark the existing method as deprecated (and maybe make it `panic!` for good measure). Then a `./x.py test` should shake out all the callers. =)

Reply:
In any case, I think we can also (in this PR) just call `dedup`. We probably don't need a *stable* ordering, either.

Reply:
In that case, we don't need to put the auto traits into the ExistentialPredicate before sorting and deduping, so can just sort and dedup the returned auto traits. Which makes this change ultimately more readable than the whole wrapping and then comparing.

--

I'll tackle the refactor in a separate PR, since it has nothing to do with this one. Due to the argument count difference, anybody calling the inherent method will get an argument count error at compile time, so I don't think the old name needs to be deprecated.